clippy_lints/src/large_future.rs

```diff
@@ -0,0 +1,78 @@
+use clippy_utils::{diagnostics::span_lint, ty::implements_trait};
+use rustc_hir::{Expr, ExprKind, LangItem, MatchSource, QPath};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_tool_lint, impl_lint_pass};
+use rustc_target::abi::Size;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// It checks for the size of a `Future` created by `async fn` or `async {}`.
+    ///
+    /// ### Why is this bad?
+    /// Due to the current [unideal implemention](https://github.com/rust-lang/rust/issues/69826) of `Generator`,
+    /// large size of a `Future` may cause stack overflows.
+    ///
+    /// ### Example
+    /// ```rust
+    /// async fn wait(f: impl std::future::Future<Output = ()>) {}
+    ///
+    /// async fn big_fut(arg: [u8; 1024]) {}
+    ///
+    /// pub async fn test() {
+    ///     let fut = big_fut([0u8; 1024]);
+    ///     wait(fut).await;
+    /// }
+    /// ```
+    ///
+    /// `Box::pin` the big future instead.
+    ///
+    /// ```txt
+    /// let fut = Box::pin(big_fut([0u8; 1024]));
+    /// ```
+    #[clippy::version = "1.68.0"]
+    pub LARGE_FUTURE,
+    nursery,
```

Comment:
The nursery group is for lints, which are currently under active development or have a high number of false positives. For this one, I would say that either `clippy::perf` or `clippy::pedantic` would be correct. I could see this generating quite many warnings in some projects, which might be harder to fix due to the type change. Having it allow-by-default in the `clippy::pedantic` is probably the safe option :)